tests/test_http_request.py

```diff
@@ -532,6 +532,60 @@ def test_from_response_formname_errors_formnumber(self):
         self.assertRaises(IndexError, self.request_class.from_response, \
                           response, formname="form3", formnumber=2)
 
+    def test_from_response_formid_exists(self):
+        response = _buildresponse(
+            """<form action="post.php" method="POST">
+            <input type="hidden" name="one" value="1">
+            <input type="hidden" name="two" value="2">
+            </form>
+            <form id="form2" action="post.php" method="POST">
+            <input type="hidden" name="three" value="3">
+            <input type="hidden" name="four" value="4">
+            </form>""")
+        r1 = self.request_class.from_response(response, formid="form2")
+        self.assertEqual(r1.method, 'POST')
+        fs = _qs(r1)
+        self.assertEqual(fs, {'four': ['4'], 'three': ['3']})
+
+    def test_from_response_formid_notexists_fallback_formname(self):
+        response = _buildresponse(
+            """<form action="post.php" method="POST">
+            <input type="hidden" name="one" value="1">
+            <input type="hidden" name="two" value="2">
+            </form>
+            <form id="form2" name="form2" action="post.php" method="POST">
+            <input type="hidden" name="three" value="3">
+            <input type="hidden" name="four" value="4">
+            </form>""")
+        r1 = self.request_class.from_response(response, formid="form3", formname="form2")
+        self.assertEqual(r1.method, 'POST')
+        fs = _qs(r1)
+        self.assertEqual(fs, {'four': ['4'], 'three': ['3']})
+
+    def test_from_response_formid_notexist(self):
+        response = _buildresponse(
+            """<form id="form1" action="post.php" method="POST">
+            <input type="hidden" name="one" value="1">
+            </form>
+            <form id="form2" action="post.php" method="POST">
+            <input type="hidden" name="two" value="2">
+            </form>""")
+        r1 = self.request_class.from_response(response, formid="form3")
+        self.assertEqual(r1.method, 'POST')
+        fs = _qs(r1)
+        self.assertEqual(fs, {'one': ['1']})
```

Comment:
short answer will be because that is what this test is supposed to do

For in-depth check the code and signature below : 
```python
# from_response method
@classmethod
    def from_response(cls, response, formname=None, formid=None, formnumber=0, formdata=None,
                      clickdata=None, dont_click=False, formxpath=None, **kwargs):
        kwargs.setdefault('encoding', response.encoding)
        form = _get_form(response, formname, formid, formnumber, formxpath)
        formdata = _get_inputs(form, formdata, dont_click, clickdata, response)
        url = _get_form_url(form, kwargs.pop('url', None))
        method = kwargs.pop('method', form.method)
        return cls(url=url, method=method, formdata=formdata, **kwargs)

# inside _get_form method
def _get_form(response, formname, formid, formnumber, formxpath):
    """Find the form element """
if formnumber is not None:
        try:
            form = forms[formnumber]
        except IndexError:
            raise IndexError("Form number %d not found in %s" %
                             (formnumber, response))
        else:
            return form
```

as you can see in `form.py` by default `from_response` assigns the value of `* 0 to  formnumber*` 
so here selection of the form by **formnumber** is like last resolve , a fall-back mechanism to prevent the code from failing if scrapy could not find any form element using the passed parameter i.e. **formname** , **formid** and **formxpath** 

now since the form element is present in object , it will select the first form element available ( index - 0) and return it and then will call the submit method of that form...

the fall back sequence will be 
* first check for formname , 
* if not present go for *formid*
* if not present go for formxpath 
* if not present go for formnumber....
>if we will pass formnumber it will select that number of from or else select 1st form (0th index)

as for this unit test ...we are checking that the scrapy indeed return the first form when passed formid parameter does not exists else this test will fail